### ISSUE TITLE
Dual Monitor Issues with Monitor #2

### ISSUE THREAD

**Issue by phlyx** [jonaskohl/CapsLockIndicator#90]
We run dual monitors and when the mouse pointer is active over the primary monitor the caps lock indicator displays properly on it, but when we move the mouse pointer to the secondary monitor the caps lock indicator turns off.  Unfortunately it’s the program we’re running on the secondary monitor that we really wanted to know if caps lock is on for so when we're in it we don't get the indicator at all.   Any way to move the indicator the secondary monitor or at the least, have the indicator stay ON when we move the mouse pointer to the secondary monitor?

**Comment by jonaskohl** [comment]
This sounds like a bug. The indicator should always show up on the monitor which contains the mouse pointer. Is the app on your second screen running in fullscreen or as topmost?

**Comment by phlyx** [comment]
Maybe was a bug in a previous version as I just downloaded the latest version and it works fine.  The indicator jumps to whichever monitor I have my mouse pointer on.

**Comment by phlyx** [comment]
Okay, not as working fine as I thought.  We use dual monitors and one is almost always running Autodesk Inventor maximized.  CapsLockIndicator works fine but then we can open a second session on the same monitor and at some point CapsLockIndicator is on the display but behind the Inventor windows.  Tried everything we could but it was only working in the background.  Have to exit the app and restart it and then it's in front of the Inventor windows again.  Still a little buggy.

**Comment by phlyx** [comment]
Had this several times where I have multiple sessions of Inventor maximized on my second display and the indicator will be behind the windows.  Have to shut down the app and restart and then it's back in front of everything but no matter what I do with the windows it stays behind until I restart.

**Comment by jonaskohl** [comment]
I will take a closer look the next few days. Unfortunately, I don't have a lot of time at my hands at the moment because of my day job. However, I will try to fix this issue as soon as possible

**Comment by phlyx** [comment]
Not a biggie, as long as I have a way around it and if all I have to do is exit, then double click on the icon, it’s easy.

Thanks,
Scott



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by phlyx** [comment]
Also wanted to pass along this isn’t just Inventor.  Once the indicator on secondary monitor ends up behind a window, it’s behind ALL windows including Office apps.  Exit and restart brings it back in front of all apps.

Thanks,
Scott



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
From: Scott Veix ***@***.***>
Sent: Tuesday, September 28, 2021 3:14 PM
To: jonaskohl/CapsLockIndicator ***@***.***>; jonaskohl/CapsLockIndicator ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [jonaskohl/CapsLockIndicator] Dual Monitor Issues with Monitor #2 (#90)

Not a biggie, as long as I have a way around it and if all I have to do is exit, then double click on the icon, it’s easy.

Thanks,
Scott



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by Yetoo1** [comment]
@phlyx Just to see if this isn't a windows bug with secondary monitor and for data in general, there's a preinstalled tool called Magnifier and it stays on top of all windows. Move the Magnifier window to the secondary monitor and/or do tests with Autodesk Inventor and see if Magnifier will overlap it. (Opening Magnifier for the first time may be zoomed in, don't move mouse around and just decrease the magnification.) Tell us if it stays in front or goes behind.

I think something can be done with SetWindowPos and WS_EX_TOPMOST flag, but I explain some potentially precluding factors later @jonaskohl. I tested to see if the indicator window will supersedes the Magnifier window and it doesn't. It seems like the CreateParams override (https://github.com/jonaskohl/CapsLockIndicator/blob/bdc3131ffac004e99eac75c7e7d07e4feceee387/CapsLockIndicatorV3/IndicatorOverlay.cs#L82) should be doing this every time Show() is called, but maybe it's only being set once and there's another way to call it. I tried calling SetWindowPos whenever IndicatorOverlay is called (see for how to implement SetWindowPos https://www.c-sharpcorner.com/uploadfile/kirtan007/make-form-stay-always-on-top-of-every-window/), but the indicator window also doesn't supersede the magnifier window. I found a Microsoft sample program that is a magnifier tool and the indicator popup also doesn't supersede it (Magnifier tool can be found here https://github.com/microsoft/Windows-classic-samples/tree/main/Samples/Magnification and the version of the tool that uses flag WS_EX_TOPMOST with SetWindowPos is in the cpp/Windowed folder (direct link to file https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/Magnification/cpp/Windowed/MagnifierSample.cpp)).

On the other hand there's this:
Interesting results occur when opening the preinstalled Magnifier and the sample code Magnifier at the same time. No matter whether I debugged the sample code Magnifier first or started the preinstalled Magnifier first, the preinstalled one always came up on top. I'm not sure if I would get different behavior if I tried starting the sample code Magnifier before the preinstalled Magnifier after a fresh reboot. 

I tried opening two instances of the sample code Magnifier and put them on each other and the windows flicker like they are competing calling the SetWindowsPos with WS_EX_TOPMOST flag each time they update. I can't do the same with the preinstalled Magnifier because it just refocuses to the already open preinstalled Magnifier window. 

I open the preinstalled Steps Recorder and hit record (hitting record causes the steps recorder to stay above windows) and open the preinstalled Magnifier and put them on each other. The preinstalled Magnifier and preinstalled Steps Recorder behave as if they were regular windows but staying on top of everything. The preinstalled Steps Recorder stays over the sample code Magnifier just like when the preinstalled Magnifier was over the sample code Magnifier and, again, no matter which order launched. 

So some possibilities: 
1. Microsoft is using an undocumented function or property to make sure the window has precedence. 
2. There is something in the background that is calculating precedence of windows that want to be TOPMOST/first layer and I/we haven't found it yet.
3. Microsoft solved the flickering/competing update issue someway that's not 1 and 2 and may be documented.
4. Bringing the window to front on an update loop is the way to go, but the Magnification sample code just isn't passing the right flags to SetWindowPos or just does this poorly and that's why it flickers when its on its self and doesn't behave like a normal window like preinstalled Magnifier and Steps Recorder do when over each other.

I am very interested in what behavior @phlyx would see if the preinstalled Magnifier and Steps Recorder were started before Autodesk Inventor on the secondary monitor and if Magnifier goes behind inventor or stays in front, because if the Autodesk guys figured this out I would expect it to act like Step Recorder and Magnifier on top of each other and go behind Autodesk Inventor. If they didn't or choose not to interfere with Magnifier because of accessibility concerns, then it would stay in front of Autodesk Inventor.

So in short, CapsLockIndicator may not be showing up because the rest of the windows are being brought to front in an update cycle faster than CapsLockIndicator assuming they are using an update cycle and not just some method or property that's better. If indeed bringing to front in an update cycle is the way to go and the sample code Magnifier is just doing that part wrong, then in terms of what needs to be changed with CapsLockIndicator, there needs to be a guarantee that the window is being set to front for every update tick and if CreateParams override is at least bringing the window to front at whatever ticks the CreateParams override covers.

**Comment by Yetoo1** [comment]
@phlyx When the indicator gets stuck behind windows on the second monitor, does it get stuck behind windows on the primary monitor when you move your cursor to the primary window and press Caps lock?

**Comment by phlyx** [comment]
I opened Magnifier and it did work for both monitors and for Inventor.

***@***.***

I am watching for what it is I do when the indicator no longer is in front.  If I stumble across something repeatable or think I know I’ll pass that along, too.

Thanks,
Scott




--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by phlyx** [comment]
I will keep an eye on things but when it gets stuck on the secondary monitor I believe it still is on top on the primary.


--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by phlyx** [comment]
Was working in Inventor on secondary, was marking up PDF’s in Foxit and running Outlook, Teams, Word, and a few other apps on primary.  Switched from primary to secondary and didn’t see the indicator pop up.  So I went back to the primary and dragged some windows up and indicator was in the background on BOTH monitors, every app was over top of them.  Exited and re-started and back to being on top.  Tried to reproduce the things I was doing but could not get it to go back to the background so unsure what triggered it.

Thanks,
Scott



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
From: Scott Veix ***@***.***>
Sent: Friday, October 1, 2021 9:56 AM
To: jonaskohl/CapsLockIndicator ***@***.***>; jonaskohl/CapsLockIndicator ***@***.***>
Cc: Mention ***@***.***>
Subject: RE: [jonaskohl/CapsLockIndicator] Dual Monitor Issues with Monitor #2 (#90)

I will keep an eye on things but when it gets stuck on the secondary monitor I believe it still is on top on the primary.


--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by phlyx** [comment]
Just had it happen again and just happened to notice the indicator wasn’t on my secondary screen.  Was doing some updates so I could reproduce what I was doing but after exiting and restarting it wouldn’t repeat.  Not sure what triggered it but just seemed sort of random.

Thanks,
Scott



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
From: Scott Veix ***@***.***>
Sent: Friday, October 1, 2021 2:14 PM
To: jonaskohl/CapsLockIndicator ***@***.***>; jonaskohl/CapsLockIndicator ***@***.***>
Cc: Mention ***@***.***>
Subject: RE: [jonaskohl/CapsLockIndicator] Dual Monitor Issues with Monitor #2 (#90)

Was working in Inventor on secondary, was marking up PDF’s in Foxit and running Outlook, Teams, Word, and a few other apps on primary.  Switched from primary to secondary and didn’t see the indicator pop up.  So I went back to the primary and dragged some windows up and indicator was in the background on BOTH monitors, every app was over top of them.  Exited and re-started and back to being on top.  Tried to reproduce the things I was doing but could not get it to go back to the background so unsure what triggered it.

Thanks,
Scott



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
From: Scott Veix ***@***.******@***.***>>
Sent: Friday, October 1, 2021 9:56 AM
To: jonaskohl/CapsLockIndicator ***@***.******@***.***>>; jonaskohl/CapsLockIndicator ***@***.******@***.***>>
Cc: Mention ***@***.******@***.***>>
Subject: RE: [jonaskohl/CapsLockIndicator] Dual Monitor Issues with Monitor #2 (#90)

I will keep an eye on things but when it gets stuck on the secondary monitor I believe it still is on top on the primary.


--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by Yetoo1** [comment]
@phlyx What os are you using and what are you're specs? Microsoft released an update for an issue relating SetWindowPos with TOPMOST flag/Window.TopMost  property, but it is for Windows 7 and seems specific to a .NET issue around siverlight (https://support.microsoft.com/en-us/topic/topmost-windows-are-not-always-in-the-topmost-position-in-windows-7-or-in-windows-server-2008-r2-24fb5117-0e97-da0a-874f-ef95122ef7cb) It seems the link to the actual download is missing, but I'll find it.

**Comment by phlyx** [comment]
Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz   3.60 GHz
64.0GB RAM
64-bit operating system, x64-based processor
Windows 10 Pro Version 21H1 (9/20/2021)
OS: 19043.1237
Windows Feature Experience Pack 120.2212.3530.0



--
Scott Veix | Sr. Mechanical Designer
RND Automation – Robotics | Packaging | Assembly
3155 Lakewood Ranch Blvd. Suite 110, Lakewood Ranch, FL 34211
941.870.5400x123
http://www.RNDautomation.com
From Concept to Reality - Custom is Standard

**Comment by phlyx** [comment]
This issue may not have much chatter BUT it is still an issue.  Have been monitoring caps lock indicator and although pretty much daily it slips to the background (behind open app windows) and I have to restart it, I have yet to catch what is specifically done that causes it to do that.

**Comment by Marin37** [comment]
While in the meantime, would an option be to add a "Refresh" or "Restart" option in the context menu between "Show" and "Exit"? Or maybe some _assignable_ keystroke/key-combo to restart it (assignable for convenience, i'm pretty sure in Inventor they're using F-keys but not which ones). Not sure if possible, but maybe something like `Ctrl`+`Alt`+`CapsLock`?

I have a similar issue (1 monitor) that sometimes it goes behind everything (not quite documented for reproduceability - also just updated, not sure if it still happens) and I solve it by restarting the program via Windows Searchbar, previously making a shortcut to the `.exe` and then copying over the Start Menu folder so that I'll press `Win` and search it directly.
